adam/visualization/positioning.py

```diff
@@ -33,86 +33,156 @@ def __init__(self, main_point: ndarray, points: List[ndarray]) -> None:
+        )
+        # self.center = torch.zeros(3, dtype=dtype, device=device)
+        self.center.requires_grad_(True)
+        self.center = nn.Parameter(self.center)
+        print(self.center)
+
+        self.corner_matrix = torch.tensor(
+            [
+                [-1, -1, -1],
+                [1, -1, -1],
+                [-1, 1, -1],
+                [1, 1, -1],
+                [-1, -1, 1],
+                [1, -1, 1],
+                [-1, 1, 1],
+                [1, 1, 1],
+            ],
+            dtype=dtype,
+        )
+        self.minkowski()
+
+    def minkowski(self):
+        # TODO: see if requires_grad=True is needed for each of these intermediate tensors
+        print(f"center: {self.center}")
+        # get all corners for current center
+        corners = torch.ones((8, 3)) * self.center + self.corner_matrix
+        print(f"corners.shape: {corners.shape}")
+        # project corners onto faces from static bb (output shape (3, 8, 1) )
+
+        projections = torch.tensor(
+            [corners[j].dot(self.static_bb_faces[i]) for i in range(3) for j in range(8)],
+            dtype=torch.float,
+        ).reshape((3, 8, 1))
+        print(f"corner projections shape: {projections.shape}")
+        print(f"corner projections: {projections}")
+
+        face_min_max = torch.tensor(
+            [[torch.min(projections[i]), torch.max(projections[i])] for i in range(3)],
+            dtype=torch.float,
+        )
+        print(f"face min-maxes shape: {face_min_max.shape}")  # (3, 2)
+        print(f"face min-maxes: {face_min_max}")
+
+        # compute overlap between the parameter's min/maxes and the static ones
+
+        overlap_ranges = torch.tensor(
+            [
+                [
+                    torch.max(
+                        face_min_max[i][0], self.static_bb_min_max_projections[i][0]
+                    ),
+                    torch.min(
+                        face_min_max[i][1], self.static_bb_min_max_projections[i][1]
+                    ),
+                ]
+                for i in range(3)
+            ],
+            dtype=torch.float,
+        )
+
+        print(f"shape of overlap ranges: {overlap_ranges.shape}")
+        print(f" overlap ranges: {overlap_ranges}")
+
+        # condense this down in to a scalar for each of the 3 (x, y, z) dimensions:
+        # to represent the degree of overlap / separation
+
+        overlap_distance = torch.tensor(
+            [overlap_ranges[i][0] - overlap_ranges[i][1] for i in range(3)],
+            dtype=torch.float,
+        )
+
+        print(f"shape of overlap distance: {overlap_distance.shape}")  # should be (3, 1)
+        print(f"overlap distance: {overlap_distance}")
+
+        # if ANY element in overlap_distance is positive, the minimum positive value
+        # is the separation distance
+
+        # otherwise the penetration distance is the maximum negative value
+        # (the smallest translation that would disentangle the two
```

Comment:
Here is the tricky part -- I'm not sure how to express the `if/else` sort of criteria here.